### PR TITLE
#4460, fix order by string column exception

### LINKED ISSUE #4460: SQL exec `order by` and `desc` feedback result may hung
Sharding-proxy version: 4.0.0-RC3

- Problems encountered 1：Query SQL `select association,rounds,election_time as electionTime,title from zhyxh where doctor_id = 848224 order by association,election_time desc` feedback results hung. 

    Query SQL `select association,rounds,election_time as electionTime,title from zhyxh where doctor_id = 848224` feedback results ok.

  logs:  no error in logs

- Problems encountered 2：Query SQL `select rtitle_ch,rn_ch,r_data,rstatus_ch,rtype_ch,researcher_ch,researcher_phone,researcher_email,researcher_addr_ch,researcher_unit_ch,exp_org_ch,applicant_ch,applicant_unit_ch,study_disease_ch from chictr where id in
                 (  
                        '2443'
                 , 
                        '2615'
                 ) 
                order by r_data desc` feedback results error `10002 - 2Unknown exception: [No signature of method: java.lang.String.mod() is applicable for argument types: (java.lang.Integer) values: [2]`.

  logs: `[DEBUG] 16:12:04.484 [ShardingSphere-Command-3] o.a.s.s.f.m.c.MySQLCommandExecutorFactory - Execute packet type: COM_QUERY, value: MySQLComQueryPacket(sql=select rtitle_ch,rn_ch,r_data,rstatus_ch,rtype_ch,researcher_ch,researcher_phone,researcher_email,researcher_addr_ch,researcher_unit_ch,exp_org_ch,applicant_ch,applicant_unit_ch,study_disease_ch from chictr where id in
                 (  
                        '2443'
                 , 
                        '2615'
                 ) 
                order by r_data desc)
[ERROR] 16:12:04.536 [ShardingSphere-Command-3] o.a.s.s.f.c.CommandExecutorTask - Exception occur: 
groovy.lang.MissingMethodException: No signature of method: java.lang.String.mod() is applicable for argument types: (java.lang.Integer) values: [2]
Possible solutions: drop(int), find(), any(), take(int), use([Ljava.lang.Object;), find(groovy.lang.Closure)
        at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:58)
        at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:49)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:58)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
        at Script109$_run_closure1.doCall(Script109.groovy:1)
        at sun.reflect.GeneratedMethodAccessor28.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1019)
        at groovy.lang.Closure.call(Closure.java:426)
        at groovy.lang.Closure.call(Closure.java:420)
        at org.apache.shardingsphere.core.strategy.route.inline.InlineShardingStrategy.execute(InlineShardingStrategy.java:90)
        at org.apache.shardingsphere.core.strategy.route.inline.InlineShardingStrategy.doSharding(InlineShardingStrategy.java:72)
        at org.apache.shardingsphere.core.strategy.route.inline.InlineShardingStrategy.doSharding(InlineShardingStrategy.java:59)
        at org.apache.shardingsphere.core.route.type.standard.StandardRoutingEngine.routeDataSources(StandardRoutingEngine.java:200)
        at org.apache.shardingsphere.core.route.type.standard.StandardRoutingEngine.route0(StandardRoutingEngine.java:188)
        at org.apache.shardingsphere.core.route.type.standard.StandardRoutingEngine.routeByShardingConditionsWithCondition(StandardRoutingEngine.java:118)
        at org.apache.shardingsphere.core.route.type.standard.StandardRoutingEngine.routeByShardingConditions(StandardRoutingEngine.java:112)
        at org.apache.shardingsphere.core.route.type.standard.StandardRoutingEngine.getDataNodes(StandardRoutingEngine.java:97)
        at org.apache.shardingsphere.core.route.type.standard.StandardRoutingEngine.route(StandardRoutingEngine.java:75)
        at org.apache.shardingsphere.core.route.router.sharding.ShardingRouter.route(ShardingRouter.java:107)
        at org.apache.shardingsphere.core.route.StatementRoutingEngine.route(StatementRoutingEngine.java:54)
        at org.apache.shardingsphere.core.SimpleQueryShardingEngine.route(SimpleQueryShardingEngine.java:58)
        at org.apache.shardingsphere.core.BaseShardingEngine.executeRoute(BaseShardingEngine.java:91)
        at org.apache.shardingsphere.core.BaseShardingEngine.shard(BaseShardingEngine.java:74)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.wrapper.StatementExecutorWrapper.doShardingRoute(StatementExecutorWrapper.java:78)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.wrapper.StatementExecutorWrapper.route(StatementExecutorWrapper.java:64)
        at org.apache.shardingsphere.shardingproxy.backend.communication.jdbc.JDBCDatabaseCommunicationEngine.execute(JDBCDatabaseCommunicationEngine.java:77)
        at org.apache.shardingsphere.shardingproxy.backend.text.query.QueryBackendHandler.execute(QueryBackendHandler.java:54)
        at org.apache.shardingsphere.shardingproxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.execute(MySQLComQueryPacketExecutor.java:73)
        at org.apache.shardingsphere.shardingproxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:86)
        at org.apache.shardingsphere.shardingproxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:66)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)`

### PR DESCRIPTION
**Pull request by tuohai666**
Fixes #4460.

### CODE REVIEW

File: sharding-core/sharding-core-merge/src/main/java/org/apache/shardingsphere/sharding/merge/dql/orderby/OrderByValue.java [hunk 1]
```diff
@@ -43,17 +52,38 @@
...
+                    }
+                } else if (orderByItemSegment instanceof IndexOrderByItemSegment) {
+                    int columnIndex = ((IndexOrderByItemSegment) orderByItemSegment).getColumnIndex();
+                    if (columns.values().toArray().length >= columnIndex) {
+                        result.add(((ColumnMetaData) columns.values().toArray()[columnIndex - 1]).isCaseSensitive());
+                        break;
+                    }
+                } else {
+                    result.add(false);
+                    break;
```
Comment (terrymanu) [inline sharding-core/sharding-core-merge/src/main/java/org/apache/shardingsphere/sharding/merge/dql/orderby/OrderByValue.java]: This method are pretty long, and lots of break in this method. Can you consider about extract them as serval small methods?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaDataLoader.java [hunk 2]
```diff
@@ -60,9 +60,34 @@
...
+            delimiterLeft = "\"";
+            delimiterRight = "\"";
+        } else if (url.startsWith("jdbc:sqlserver:")) {
+            delimiterLeft = "[";
+            delimiterRight = "]";
+        } else {
+            delimiterLeft = "";
+            delimiterRight = "";
+        }
+        return "SELECT * FROM " + delimiterLeft + table + delimiterRight + " WHERE 1 != 1;";
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaDataLoader.java]: Please re-design about the codes here.
We should never do something about the sql dialect in this module.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaData.java [hunk 3]
```diff
@@ -36,4 +37,7 @@
     private final String dataType;
     
     private final boolean primaryKey;
+    
+    @Setter
+    private boolean isCaseSensitive;
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaData.java]: Please do not use setter, we may keep consist with other fields.

Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaData.java]: Please consider change `isCaseSensitive` to `caseSensitive`

File: sharding-core/sharding-core-merge/src/main/java/org/apache/shardingsphere/sharding/merge/dql/orderby/OrderByValue.java (hunk 1 above)
Comment (tuohai666) [reply sharding-core/sharding-core-merge/src/main/java/org/apache/shardingsphere/sharding/merge/dql/orderby/OrderByValue.java]: Fixed

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaData.java (hunk 3 above)
Comment (tuohai666) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaData.java]: Fixed.

Comment (tuohai666) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaData.java]: Fixed.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaDataLoader.java (hunk 2 above)
Comment (tuohai666) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/metadata/column/ColumnMetaDataLoader.java]: Fixed.

**Pull request merged**